src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include <workerd/jsg/filesystem.h>
+#include <workerd/jsg/url.h>
+
+// Every worker instance will have its own root directory (/). In this root
+// directory we will have at least two special directories, the "bundle" root
+// and the "temp" root. More special directories can be added later.
+//
+// The bundle root is where all of the modules that are included in the worker
+// bundle will be accessible. Everything in this directory will be strictly
+// read-only. The contents are populated by the worker configuration bundle.
+// By default, the bundle root will be /bundle but this can be overridden
+// using the FsMap API defined below.
+//
+// The temp root is where we will allow temporary files to be created.
+// Everything in this directory is read-write but will be transient. By default,
+// the temp root will be /tmp but this can also be overridden.
+//
+// Let's imagine the following simple workerd configuration:
+//
+// ```
+// const helloWorld :Workerd.Worker = (
+//   modules = [
+//     (name = "worker",
+//      esModule = embed "worker.js"),
+//     (name = "foo", text = "Hello World!"),
+//   ],
+//   compatibilityDate = "2023-02-28",
+// );
+// ```
+//
+// Given this configuration, the worker fs will initially have the following structure:
+//
+// /
+// ├── bundle
+// │   ├── worker
+// │   └── foo
+// └── tmp
+//
+// We can then use the jsg::Lock& to access the file system at the C++ level.
+//
+// ```cpp
+// jsg::Lock& js = ...
+// // Resolve the root of the file system
+// KJ_IF_SOME(node,
+//            js.resolveVfsNode("file:///"_url)) {
+//   KJ_SWITCH_ONEOF(node) {
+//     KJ_CASE_ONEOF(file, kj::Own<jsg::fs::File>) {
+//       // ...
+//     }
+//     KJ_CASE_ONEOF(dir, kj::Own<jsg::fs::Dir>) {
+//       // ...
+//     }
+//   }
+// }
+// ```
+//
+// The temporary file directory is a bit special in that the contents are fully
+// transient based on whether there is an active IoContext or not. We use a
+// special RAII TempDirStoreScope scope to manage the contents of the temporary
+// directory. For example,
+//
+// ```cpp
+// KJ_IF_SOME(node,
+//     js.resolveVfsNode("file:///tmp"_url)) {
+//   auto& dir = KJ_ASSERT_NONNULL(node.tryGet<kj::Own<jsg::fs::Directory>>());
+//   TempDirStoreScope temp_dir_scope;
+//   kj::Path path("a/b/c/foo.txt")
+//   auto tmpFile = dir.tryOpen(js, path, jsg::fs::Node::FILE);
+//   KJ_ASSERT(tmpFile.write(js, 0, "Hello World!"_kjb) == 12);
```

Comment:
Can you add a comment about write() function returning the length of the input.

Reply:
In this case it does. What it actually returns is the number of bytes written, which depending on the underlying implementation may be <= length of input.